timeinterval/repeating_interval.go

```diff
@@ -0,0 +1,164 @@
+	Interval Interval
+	RepeatIn time.Duration
+	Repetitions *uint32
+}
+
+// UnmarshalJSON unmarshal RepeatingInterval from an ISO8601 "repeating interval" string.
+func (in *RepeatingInterval) UnmarshalJSON(data []byte) error {
+	var s string
+	err := json.Unmarshal(data, &s)
+	if err != nil {
+		return err
+	}
+	ri, err := ParseRepeatingIntervalISO8601(s)
+	if err != nil {
+		return err
+	}
+	*in = *ri
+	return nil
+}
+
+// MarshalJSON marshal RepeatingInterval into an ISO8601 "repeating interval" string.
+func (in RepeatingInterval) MarshalJSON() ([]byte, error) {
+	s, err := in.ISO8601()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(s)
+}
+
+// StartsAt returns the time the interval begins.
+// When possible StartsAt will be derived using the Duration and Repetitions fields if Interval.StartsAt is unset.
+func (in RepeatingInterval) StartsAt() *time.Time {
+	if in.isStartsAtBoundedByRepetitions() {
+		startsAt := in.Interval.EndsAt().Add(-time.Duration(*in.Repetitions)  * in.RepeatIn)
+		return &startsAt
+	}
+	return in.Interval.StartsAt()
+}
+
+// EndsAt returns the time the interval ends.
+// When possible EndsAt will be derived using the Duration and Repetitions fields if Interval.EndsAt is unset.
+func (in RepeatingInterval) EndsAt() *time.Time {
+	if in.isEndsAtBoundedByRepetitions() {
+		endsAt := in.Interval.StartsAt().Add(time.Duration(*in.Repetitions)  * in.RepeatIn)
+		return &endsAt
+	}
+	return in.Interval.EndsAt()
+}
+
+// Duration returns the duration the repeating interval will be active for or nil if it is unbounded.
+func (in RepeatingInterval) Duration() *time.Duration {
+	endsAt := in.EndsAt()
+	startsAt := in.StartsAt()
+	if startsAt == nil || endsAt == nil {
+		return nil
+	}
+	d := endsAt.Sub(*startsAt)
+	return &d
+}
+
+// Started returns a boolean indicating if the interval has begun at the given time.
+func (in RepeatingInterval) Started(t time.Time) bool {
+	if in.isStartsAtBoundedByRepetitions() {
+		startsAt := in.StartsAt()
+		return t.Equal(*startsAt) || t.After(*startsAt)
+	}
+	return in.Interval.Started(t)
+}
+
+// Ended returns a boolean indicating if the interval has ended at the given time.
+func (in RepeatingInterval) Ended(t time.Time) bool {
+	if in.isEndsAtBoundedByRepetitions() {
+		endsAt := in.EndsAt()
+		return t.After(*endsAt)
+	}
+	return in.Interval.Ended(t)
+}
+
+// In returns a boolean indicating if the given time is when the interval is active (Started and not Ended)
+func (in RepeatingInterval) In(t time.Time) bool {
+	return in.Started(t) && !in.Ended(t)
+}
+
+// Next returns the time of the next interval-occurrence relative to the given time.
+// It returns the startsAt time if the interval have not started yet and nil if the interval has ended.
+func (in RepeatingInterval) Next(t time.Time) *time.Time {
+	if !in.Started(t) {
+		return in.StartsAt()
+	}
+	if in.Ended(t) || in.RepeatIn == 0 {
+		return nil
+	}
+	startsAt := in.StartsAt()
+	diff := t.Sub(*startsAt)
```

Comment:
Crash if startsAt is `nil`.

Reply:
It actually won't, there also is a test proving this :)

Ref: https://github.com/corthmann/go-time-intervals/blob/master/timeinterval/repeating_interval_test.go#L38